### PR TITLE
Persiting effective permissions and relations

### PR DESCRIPTION
**Pull request by daniel-jodlos**

### CODE REVIEW

File: src/events/event_processor.erl [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+-spec async_process(binary(), map()) -> pid().
+async_process(Vertex, Event) ->
+    spawn(?MODULE, process, [Vertex, Event]).
+
+-spec process(binary(), map()) -> ok | {error, any()}.
+process(Vertex, Event) ->
+    instrumentation:event_started(Event),
+    #{<<"type">> := Type, <<"trace">> := Trace, <<"sender">> := Sender,
+        <<"originalSender">> := OriginalSender, <<"effectiveVertices">> := EffectiveVertices} = Event,
+    EventTuple = {Type, Trace, Sender, OriginalSender, EffectiveVertices},
```
Comment (MrPaulMarshall) [inline src/events/event_processor.erl:43]: This was just a quick fix for lint.
However, the values in this tuple DON'T actually compliant with type `event()`

So, it has to be resolved, one way or another

**Comment by daniel-jodlos** [comment]
Musimy to zmergować raczej wcześniej niż później, także może zostawmy refactor na później i skupmy się tylko na logice

File: src/rest/handlers/rest_queries.erl [hunk 2]
```diff
@@ -255,17 +255,19 @@ members_naive_locally(Of, JumpCount) ->
...
+        {ok, #{<<"permissions">> := Permissions}} -> {ok, Permissions};
+        {error, Reason} -> {error, Reason}
+    end.
 
-%% @todo
 -spec members_indexed(binary()) -> {ok, list(binary())} | {error, any()}.
-members_indexed(_Of) ->
-    {ok, []}.
+members_indexed(Of) ->
+    % {ok, graph:effective_list_children(Of) ++ graph:effective_list_parents(Of)}.
```
Comment (MrPaulMarshall) [inline src/rest/handlers/rest_queries.erl:272]: Remove this line.

File: src/graph.erl [hunk 3]
```diff
@@ -297,8 +409,66 @@ list_parents(Vertex) -> persistence:set_list_members(parents_id(Vertex)).
...
+    timer:sleep(1000),
+    io:format("Next test case~n"),
+    create_edge(E, A, Perms),
+    timer:sleep(1000),
+    io:format("Next test case~n"),
+    create_edge(C, F, Perms),
+    timer:sleep(1000),
+    io:format("Next test case~n"),
+    remove_edge(E, A),
+    ok.
```
Comment (MrPaulMarshall) [inline src/graph.erl:474]: Why is it here?

File: src/graph.erl [hunk 4]
```diff
@@ -297,8 +409,66 @@ list_parents(Vertex) -> persistence:set_list_members(parents_id(Vertex)).
...
 %% @todo implement this - basically all zones that this zone has contact with
 -spec all_zones() -> {ok, list(binary())} | {error, any()}.
 all_zones() ->
     {ok, []}.
+
+-spec effective_list_parents(Vertex :: binary()) -> list(binary()).
+effective_list_parents(Vertex) -> do_list(persistence:set_list_members(effective_parents_id(Vertex))).
+
+do_list({ok, Result}) -> Result;
+do_list({error, _Error}) -> [].
```
Comment (MrPaulMarshall) [inline src/graph.erl:421]: This is incompatible with every other function in this module. Why did you do this like that?

File: src/graph.erl [hunk 5]
```diff
@@ -297,8 +409,66 @@ list_parents(Vertex) -> persistence:set_list_members(parents_id(Vertex)).
...
+
+add_effective_child(Vertex, Child) ->
+    persistence:set_add(effective_children_id(Vertex), Child).
+remove_effective_child(Vertex, Child) ->
+    persistence:set_remove(effective_children_id(Vertex), Child).
+
+-spec effective_list_children(Vertex :: binary()) -> list(binary()).
+effective_list_children(Vertex) -> do_list(persistence:set_list_members(effective_children_id(Vertex))).
+
+%% @TODO change shape of the event to be compliant with "standard" (ask Pawel for details)
```
Comment (MrPaulMarshall) [inline src/graph.erl:436]: Remove this line

File: src/graph.erl [hunk 6]
```diff
@@ -297,8 +409,66 @@ list_parents(Vertex) -> persistence:set_list_members(parents_id(Vertex)).
...
+
+post_events_about_effective_children(Vertex, TargetVertex, Type) ->
+    Children = effective_list_children(Vertex),
+    Event = #{<<"id">> => gmm_utils:uuid(), <<"type">> => <<"child/", Type/binary>>, <<"originalSender">> => Vertex, <<"sender">> => Vertex, <<"effectiveVertices">> => [Vertex | Children], <<"trace">> => <<"null">>},
```
Comment (MrPaulMarshall) [inline src/graph.erl:440]: Why is trace set to string "null"? Where the heck did it come from?

File: src/graph.erl [hunk 7]
```diff
@@ -297,8 +409,66 @@ list_parents(Vertex) -> persistence:set_list_members(parents_id(Vertex)).
...
+    inbox:post(TargetVertex, Event).
+
+post_events_about_effective_parents(Vertex, TargetVertex, Type) ->
+    Parents = effective_list_parents(Vertex),
+    Event = #{<<"id">> => gmm_utils:uuid(), <<"type">> => <<"parent/", Type/binary>>, <<"originalSender">> => Vertex, <<"sender">> => Vertex, <<"effectiveVertices">> => [Vertex | Parents], <<"trace">> => <<"null">>},
```
Comment (MrPaulMarshall) [inline src/graph.erl:445]: Again, why trace is set to "null"?

File: src/events/event_processor.erl [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+-export([
+    process/2,
+    async_process/2
+]).
+
+async_process(Vertex, Event) ->
+    spawn(?MODULE, process, [Vertex, Event]).
+    
+process(Vertex, Event) ->
+    base_instrumentation:event_started(Vertex, Event),
```
Comment (MrPaulMarshall) [inline src/events/event_processor.erl:12]: Use my ready module for instrumentation. Just delete your dummy one.

File: src/events/event_processor.erl [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    #{<<"type">> := Type} = Event,
+    case Type of
+        <<"child/updated">> -> process_child_change(Vertex, Event);
+        <<"child/removed">> -> process_child_removed(Vertex, Event);
+        <<"parent/updated">> -> process_parent_change(Vertex, Event);
+        <<"parent/removed">> -> process_parent_removed(Vertex, Event)
+    end,
+    base_instrumentation:event_finished(Vertex, Event),
+    inbox:free_vertex(Vertex),
```
Comment (MrPaulMarshall) [inline src/events/event_processor.erl:21]: That is dangerous. Wrap it in try-catch and ALWAYS free vertex, regardless of success of the operation.

File: src/events/event_processor.erl [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+calculatePermissions(From, To) ->
+    Intermediate = graph:get_intermediate_verticies(From, To, To),
+    IntermediatePermissions = lists:map(fun (Vertex) ->
+        case graph:get_edge(Vertex, To) of
+            {ok, #{<<"permissions">> := Permissions}} -> Permissions;
+            _other -> <<"00000">>
+        end
+    end, Intermediate),
+    io:format("DUPA: ~p~n", [IntermediatePermissions]),
```
Comment (MrPaulMarshall) [inline src/events/event_processor.erl:111]: Perfect

**Review by MrPaulMarshall — CHANGES_REQUESTED** [review]
You can leave style-related things for now, of course. However, there are some sections, like instrumentation module, that you should've done already.

Regarding logic: I don't see any errors (except the fact that you don't free vertex in case of an error) BUT it's late and I could've just miss something.

File: src/events/event_processor.erl (hunk 10 above)
Comment (daniel-jodlos) [reply src/events/event_processor.erl:111]: Leaving it in then!

File: src/graph.erl (hunk 4 above)
Comment (daniel-jodlos) [reply src/graph.erl:421]: In optimistic scenario, which we do plenty of here, {error, Reason} indicates that the entry doesn't exist. This function is used for effective children and parents and if there isn't an entry in database, there are no effective children and parents

Comment (MrPaulMarshall) [reply src/graph.erl:420]: I suppose. It seems the better option is that database returns an error (not an empty result) then naive query should be called.

I will accept it for now as I want to run tests on k8s, but you should refactor it asap.

**Review by MrPaulMarshall — APPROVED** [review]
I approve this to run tests on k8s, BUT you should refactor it (first and foremost convention of return values in graph.erl etc)

**Pull request merged**